### PR TITLE
Expand Parameter Validation Implicit VU Coverage

### PR DESCRIPTION
**Pull request by mark-lunarg**
This patch series:

- adds implicit VU coverage for structs _only_ appearing in pNext chains
- added VUIDs for validation of any struct or substruct whose contents are validated
- added correct VUIDs for string integrity checks
- added VUIDs for stype checks for arrays of structs

This increases implicit VU coverage by 250+ checks.  

Still around 400 or so to be picked up.

### CODE REVIEW

**Comment by mark-lunarg** [comment]
Fixed the clang-format issue locally.

File: scripts/parameter_validation_generator.py [hunk 1]
```diff
@@ -500,6 +544,16 @@ def genStruct(self, typeinfo, typeName, alias):
...
                                                 condition=conditions[name] if conditions and name in conditions else None,
                                                 cdecl=cdecl))
+        # If this struct extends another, keep its name in list for further processing
+        if (typeinfo.elem.attrib.get('structextends') is not None):
+            self.structextends_list.append(typeName)
+        # Returnedonly structs should have most of their members ignored -- on entry, we only care about validating the sType and
+        # pNext members. Everything else will be overwritten by the callee.
+        if typeinfo.elem.attrib.get('returnedonly') is not None:
+            if typeName == 'VkPhysicalDeviceGroupProperties':
+                stop = 'here'
```
Comment (tobine) [inline scripts/parameter_validation_generator.py]: what do these `stop = 'here'` lines do?

File: layers/parameter_validation.h [hunk 2]
```diff
@@ -176,6 +176,19 @@ struct LogMiscParams {
     const char *api_name;
 };
 
+// Pick up VUIDS from validated substructures. Many structures contain substructures.  In addition to the implicit VUs covered
+// when each of the substructures is parsed, there is often a VUID for 'xxx MUST be a valid XxxStruct'. If all implicit valid usage
+// is covered for a substructure, we should claim the VUID for the validity of the structure itself.
+static bool ValidateSubstructure(debug_report_data *report_data, const char *api_name, const char *struct_name,
+                                 const std::string &substructure_vuid) {
+    bool skip = false;
+    if (false) {
```
Comment (tobine) [inline layers/parameter_validation.h]: `"if (false)"`? Wazzup?

File: scripts/vuid_mapping.py [hunk 3]
```diff
@@ -1168,6 +1173,40 @@
...
 'srcOffsets' : 535,
 'dstOffsets' : 536,
+'offset' : 537,
+'extent' : 538,
+'pMapEntries' : 539,
+'color' : 540,
+'depthStencil' : 541,
+'pTessellationState' : 542,
+'pViewportState' : 543,
+'pMultisampleState' : 543,
```
Comment (tobine) [inline scripts/vuid_mapping.py]: 543 duplicated and 545 skipped.

**Comment by tobine** [comment]
Also, tests? Don't need tests for all cases but perhaps at least 4 new tests to hit one each of the four main classes of checks noted in your PR description?

File: scripts/vuid_mapping.py (hunk 3 above)
Comment (mark-lunarg) [reply scripts/vuid_mapping.py]: Thanks!

File: layers/parameter_validation.h (hunk 2 above)
Comment (mark-lunarg) [reply layers/parameter_validation.h]: So.  It was intended that this message would never actually be output to the user.  The spec is kinda weird here -- Each structure has an implicit VU that says 'the structure must be a valid VkXxx structrure', but in addition, all members of that structure get their own VUs.  We could output this message, but than any implicit check would always trigger two errors.  I thought this was better, but am open to suggestions.

File: scripts/parameter_validation_generator.py (hunk 1 above)
Comment (mark-lunarg) [reply scripts/parameter_validation_generator.py]: This is already gone...

File: layers/parameter_validation.h (hunk 2 above)
Comment (tobine) [reply layers/parameter_validation.h]: Can you give an example of two overlapping VUs that would be flagged?
For structures I always see the first implicit VU of "sType must be VK_STRUCTURE_TYPE_..."
Are the duplicate errors only a concern when a pNext chain struct has wrong type. If that's the case the conflict still isn't clear to me on just a bit of thought.
The sType and pNext errors are independent to me so I think I'm missing something here.

Comment (mark-lunarg) [reply layers/parameter_validation.h]: Sure, in CreateInstance, the first implicit VU is "**pCreateInfo must be a valid pointer to a valid VkInstanceCreateInfo structure**" (**VUID-vkCreateInstance-pCreateInfo-parameter**).  But, that same **VkInstanceCreateInfo** structure has the following implicit VUIDs:

* sType must be VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO
* Each pNext member of any structure (including this one) in the pNext chain must be either NULL or a pointer to a valid instance of VkDebugReportCallbackCreateInfoEXT, VkDebugUtilsMessengerCreateInfoEXT, or VkValidationFlagsEXT
* Each sType member in the pNext chain must be unique
* flags must be 0
* If pApplicationInfo is not NULL, pApplicationInfo must be a valid pointer to a valid VkApplicationInfo structure
* If enabledLayerCount is not 0, ppEnabledLayerNames must be a valid pointer to an array of enabledLayerCount null-terminated UTF-8 strings
* If enabledExtensionCount is not 0, ppEnabledExtensionNames must be a valid pointer to an array of enabledExtensionCount null-terminated UTF-8 strings

Theoretically, **any** failure among those checks should ALSO trigger the CreateInstance VUID failure.

Comment (tobine) [reply layers/parameter_validation.h]: Ok, that makes more sense. For the first VU my opinion is that we should just flag it if the pCreateInfo ptr is NULL. For all other cases just flag the errors for the specific struct params.

I agree that a strict adherence to VU language would require flagging the first error for any bad struct param, but if I'm a user the redundancy doesn't help me; I only need the more specific error to fix my issue.

**Comment by mark-lunarg** [comment]
@tobine, moved the fake substructure checks into an existing struct check, which now returns the substruct VUID only if the pointer is null, as you suggested.  This is also much cleaner.  Made the other changes as well, and added four new tests.

File: scripts/parameter_validation_generator.py [hunk 4]
```diff
@@ -493,6 +534,14 @@ def genStruct(self, typeinfo, typeName, alias):
                                                 extstructs=self.registry.validextensionstructs[typeName] if name == 'pNext' else None,
                                                 condition=conditions[name] if conditions and name in conditions else None,
                                                 cdecl=cdecl))
+        # If this struct extends another, keep its name in list for further processing
+        if (typeinfo.elem.attrib.get('structextends') is not None):
```
Comment (tobine) [inline scripts/parameter_validation_generator.py]: parens around non-compound "if" clause are not common in the code, see line 542 below

File: tests/layer_validation_tests.cpp [hunk 5]
```diff
@@ -1348,6 +1353,74 @@ TEST_F(VkLayerTest, RequiredParameter) {
...
+    for (auto device_extension : required_device_extensions) {
+        if (DeviceExtensionSupported(gpu(), nullptr, device_extension)) {
+            m_device_extension_names.push_back(device_extension);
+        } else {
+            printf("%s %s Extension not supported, skipping tests\n", kSkipPrefix, device_extension);
+            return;
+        }
+    }
+    // Create a device passing in a bad PdevFeatures2 value
+    auto indexingFeatures = lvl_init_struct<VkPhysicalDeviceDescriptorIndexingFeaturesEXT>();
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: camelCase slipped through

**Pull request merged**